### PR TITLE
Refactor DAG system for more robustness.

### PR DESCRIPTION
**Pull request by mmcdermott**

### CODE REVIEW

**Comment by mmcdermott** [comment]
Do not merge this yet; I just made it into a PR for ease of review.

File: src/lib/DAG/DAG.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import {orFns} from 'gEngine/collections'
+import {typeSafeEq, mutableCopy, orArr} from 'gEngine/utils'
+import * as errorTypes from 'lib/propagation/errors'
+
+const matchesId = testId => n => !!n.isCycle ? _.some(n.nodes, ({id}) => typeSafeEq(testId, id)) : typeSafeEq(testId, n.id)
```
Comment (OAGr) [inline src/lib/DAG/DAG.js]: what is n? A nodeSet?

File: src/lib/DAG/DAG.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  while (!_.isEmpty(unprocessedNodes)) {
+    const nextLevelNodes = _.remove(unprocessedNodes, allInputsWithin(_.flatten(heightOrderedNodes), missingInputs))
+    if (_.isEmpty(nextLevelNodes)) { console.warn('INFINITE LOOP DETECTED'); break }
+    heightOrderedNodes.push(nextLevelNodes)
+  }
+  return heightOrderedNodes
+}
+
+const isRelatedToFn = ({id}, ancestors) => n => ancestors[id].includes(n.id) || ancestors[n.id].includes(id)
+const anyRelationsWithin = (nodeSet, ancestors) => orFns(...nodeSet.map(n => isRelatedToFn(n, ancestors)))
```
Comment (OAGr) [inline src/lib/DAG/DAG.js]: what is the difference between a nodeSet and nodes?

Comment (mmcdermott) [inline src/lib/DAG/DAG.js]: No difference. Will fix.

File: src/lib/DAG/DAG.js (hunk 1 above)
Comment (mmcdermott) [inline src/lib/DAG/DAG.js]: Let me know if it is clearer after the next push.

File: src/lib/DAG/DAG-test.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            {id: 'I', inputs: ['H']},
+            {id: 'J', inputs: ['I']},
+            {id: 'K', inputs: ['J']}
+          ]
+        },
+      ])
+    })
+  })
+
+  describe('separateIntoHeightSets', () => {
```
Comment (OAGr) [inline src/lib/DAG/DAG-test.js]: add spaces

Comment (OAGr) [inline src/lib/DAG/DAG-test.js]: if you use method names, use # in beginning

**Comment by OAGr** [comment]
This seems like a pretty good improvement, great work here.

**Pull request merged**